### PR TITLE
Migrate to use v0.14 API

### PR DESCRIPTION
**Pull request by cosmo0920**
I've tried to use v0.14 Output plugin API.
This PR contains major update change.
Could you bump up minor version if releasing new version of gem?
And please refer http://docs.fluentd.org/v0.14/articles/plugin-update-from-v12 before release new version.

Thanks,

### CODE REVIEW

File: lib/fluent/plugin/out_memcached.rb [hunk 1]
```diff
@@ -33,12 +31,21 @@ def start
 
   def shutdown
     @memcached.close
+    super
   end
 
   def format(tag, time, record)
     [tag, time, record].to_msgpack
```
Comment (cosmo0920) [inline lib/fluent/plugin/out_memcached.rb:38]: In v0.14, we can include tag information and reduce chunk size.
Should this plugin reduce chunk size? 
@innossh How do you think of this question?

**Comment by innossh** [comment]
@cosmo0920 Thanks for the PR. I'll check it later. One question, you know the latest stable version of fluentd is v0.12. Do you want to urgently release the new version of this plugin which uses v0.14 API?

**Comment by cosmo0920** [comment]
> Do you want to urgently release the new version of this plugin which uses v0.14 API?

Sooner is better. But I don't want to urgently release this plugin.
Releasing prerelease version or merging after v0.14 get stable is also acceptable for me.

File: lib/fluent/plugin/out_memcached.rb (hunk 1 above)
Comment (innossh) [reply lib/fluent/plugin/out_memcached.rb:38]: > In v0.14, we can include tag information and reduce chunk size.

It means `<buffer>` or `<inject>` section, right? Could you give me an example of config or code to reduce chunk size? I'm sorry I don't have much knowledge of v0.14.

Comment (cosmo0920) [reply lib/fluent/plugin/out_memcached.rb:38]: > It means <buffer> or <inject> section, right?

Yes. I mean that with the following configuration, Fluentd v0.14 can contain tag information in chunk's metadata:

```aconf
<match **>
   @type memcached
   <buffer tag>
      @type memory
      # <snip>
   </buffer>
   # <snip>
</match>
```



> Could you give me an example of config or code to reduce chunk size?

The following code is an example code for reducing chunk size.

```ruby
config_section :buffer do
  config_set_default :@type, DEFAULT_BUFFER_TYPE
  # The below config enforces to contain tag information in chunk.
  config_set_default :chunk_keys, ['tag'] 
end

def configure(conf)
  super
  # <snip>
  unless @chunk_key_tag
    raise Fluent::ConfigError, "'tag' in chunk_keys is required."
  end
end

def format(tag, time, record)
  # Omit tag serialization
  [time, record].to_msgpack
end

def write(chunk)
  tag = chunk.metadata.tag
  chunk.msgpack_each do |time, record|
  # <snip>
  end
end
```

Comment (innossh) [reply lib/fluent/plugin/out_memcached.rb:38]: Thank you for the example! Now I understand. It's better to reduce the chunk size.
Just for my curiosity, what kind of cases should the chunk include tag? (It means `[tag, time, record].to_msgpack`, not `[time, record].to_msgpack`)

Comment (cosmo0920) [reply lib/fluent/plugin/out_memcached.rb:38]: > what kind of cases should the chunk include tag?

I guess that an one case has as the below config:

```aconf
<match **>
   @type memcached
   <buffer time>
      @type memory
      timekey 1h
      # <snip>
   </buffer>
   # <snip>
</match>
```

And tag is needed in `#write` case.

Comment (innossh) [reply lib/fluent/plugin/out_memcached.rb:38]: OK. Thank you for the information!

**Comment by innossh** [comment]
Thanks @cosmo0920 !

**Pull request merged**